### DISCUSSION TITLE
Sort not working when field doesn't exist. (MongoDB)

### DISCUSSION THREAD

**Discussion by Atlinx** [twinlogix/typetta#288] [Q&A] [answered]
When I sort for a field that doesn't exist on all items, the sort glitches out -- specifically, it results in duplicate elements during pagination. I think when an item doesn't have a field being sorted, that item should be placed last.

**Comment by edobrb** [comment]
Are you using the in-memory driver? (mock)

**Reply by Atlinx** [reply]
I've tested it with a real mongodb database and a mock.

**Comment by Atlinx** [comment]
Weird, everything seemed to sort itself out. Maybe it was an issue with my frontend that led to this. Might be the same reason why #287 disappeared as well for me.

**Reply by edobrb** [reply]
Thank you for the feedback!

**Comment by Atlinx** [comment, marked as answer]
@edobrb  On second thought, this issue resurfaced again. I realized that it was only fixed because I was working with a database that had the field set. When new entries came in without the field, this bug appeared again.

~~I'm going to do some investigating on my side to see if it's a frontend issue.~~

EDIT: See reply below

**Reply by Atlinx** [reply]
I think I've found the cause. MongoDB skip and limit create duplicates: https://stackoverflow.com/questions/44692075/mongo-db-duplication-issue-while-using-sorting-with-limit-and-skip-in-aggregatio. This explains why I can't reproduce this with the mock. I suggest changing the MongoDB driver to not return duplicates for the sake of consistency across all drivers.

**Reply by edobrb** [reply]
In practice, the proposed solution is to add an order by id in order to avoid duplicates. However, this will have an impact on performance even on queries for which this operation would not be required. Furthermore, the problem is also present when using the MongoDB driver. I'm not sure Typetta has to fix this. Take a look at [https://stackoverflow.com/questions/56609607/mongodb-strange-behaviour-of-sort-with-pagination-leading-to-duplicates](https://stackoverflow.com/questions/56609607/mongodb-strange-behaviour-of-sort-with-pagination-leading-to-duplicates) he suggests adding an index to the field you are ordering on.
Let me know what you think

**Reply by edobrb** [reply]
The fix would be this: [https://github.com/twinlogix/typetta/blob/6fa41fd61b5c82159beea5a5850dd42f4b7f9f6b/src/dal/drivers/no-sql/mongodb/dao.mongodb.ts#L63-L65](https://github.com/twinlogix/typetta/blob/6fa41fd61b5c82159beea5a5850dd42f4b7f9f6b/src/dal/drivers/no-sql/mongodb/dao.mongodb.ts#L63-L65)

**Reply by Atlinx** [reply]
Would I be able to leverage using the index out the box with Typetta, or would I have to write a manual MongoDB-specific query?

**Reply by Atlinx** [reply]
Ah that did the trick! Thanks! ❤️

**Reply by edobrb** [reply]
At the moment Typetta does not support automatic index creation. Btw you could use the `execQuery`:
```typescript
await entityManager.execQuery(async (dbs, entities) => {
    await entities.user?.createIndex(/* ... */)
})
```

or you could use something like https://www.npmjs.com/package/migrate-mongo for seeding initial data and creating indexes only once.
